lib/alloc.c

```diff
@@ -71,3 +71,23 @@ xreallocarray(void *p, size_t nmemb, size_t size)
 	        log_get_progname(), strerror(errno));
 	exit(13);
 }
+
+char *
+xstrndup(const char *s, size_t max_length) {
+    size_t length = strlen(s);
+    if (length > max_length) {
+        length = max_length;
+    }
+
+    char *p = malloc(length + 1);
+    if (!p) {
+		fprintf(log_get_logfd(), _("%s: %s\n"),
+	    		log_get_progname(), strerror(errno));
+        exit(13);
```

Comment:
Isn't this simpler as the following?:

```c
{
        char  *dup;

        dup = strndup(s, n);
        if (dup == NULL) { ... }

        return dup;
}
```

Reply:
Or even simpler as:


```c
{
        return strcat(memcpy(XMALLOC(strnlen(str, n) + 1, char), str, strnlen(str, n)), "");
}
```

Disclaimer: not tested.

Reply:
New approach is applied by direct usage of a new function: is_valid_user_name